packet.go

```diff
@@ -154,11 +154,18 @@ func recvPacket(r io.Reader, alloc *allocator, orderID uint32) (uint8, []byte, e
 		debug("recv packet %d bytes too long", length)
 		return 0, nil, errLongPacket
 	}
+	if length == 0 {
+		debug("recv packet of %d bytes too short", length)
+		return 0, nil, errShortPacket
+	}
 	if alloc == nil {
 		b = make([]byte, length)
 	}
 	if _, err := io.ReadFull(r, b[:length]); err != nil {
 		debug("recv packet %d bytes: err %v", length, err)
+		if err == io.EOF || err == io.ErrUnexpectedEOF {
+			err = errShortPacket
```

Comment:
I mean, technically, in these cases, I would expect `io.ErrUnexpectedEOF`.

We’re trying to read a packet of a `length > 0` but before we can read it all, we encounter an EOF. Perfect case for an `io.ErrUnexpectedEOF`. It‘s not a too short packet (where we read the whole length but an internal value has a length that is longer than the packet) it‘s a proper length packet, but like… EOF before reading it all. That‘s unexpected.

(Sorry to keep “waffling” like this. Code polishing is good, but sometimes you end up going a different direction, then the wind turns, and oh no, now you’re stuck going the wrong way.)

Reply:
I think io.ErrUnexpectedEOF is conflating two error conditions. If client code calls io.ReadFull on an sftp.File, it cannot distinguish a protocol error/server crash from the premature end of an actual file on the remote. Maybe errShortPacket isn't the right value, but there should be a stronger hint that something went wrong in transmission.

Come to think of it, the same may hold true for the ReadFull above, which can return both kinds of EOF...

Or maybe I'm misreading something and I should check again in the morning.

Reply:
I'm having a very hard time simulating this in a test, but I did notice the following. When I print the error return from io.Copy in TestServerRoughDisconnect,

https://github.com/pkg/sftp/blob/3c22ebff33f246c81ec85ef7fedb04c2bb2da018/client_integration_test.go#L1836

... that prints nil, because io.Copy does not consider EOF an error. Will it help the review for this issue if I remove this check and open a separate issue/PR for this?

Reply:
I'll open a separate issue.